libs/server-sdk/src/data_store/data_store_updater.hpp

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+#include "../data_source/data_source_update_sink.hpp"
+#include "data_store.hpp"
+#include "dependency_tracker.hpp"
+
+#include <launchdarkly/server_side/change_notifier.hpp>
+
+#include <boost/signals2/signal.hpp>
+
+#include <memory>
+
+namespace launchdarkly::server_side::data_store {
+
+class DataStoreUpdater
+    : public launchdarkly::server_side::data_source::IDataSourceUpdateSink,
+      public launchdarkly::server_side::IChangeNotifier {
+   public:
+    template <typename Storage>
+    using Collection = data_model::SDKDataSet::Collection<std::string, Storage>;
+
+    template <typename Storage>
+    using SharedItem = std::shared_ptr<data_model::ItemDescriptor<Storage>>;
+
+    template <typename Storage>
+    using SharedCollection =
+        std::unordered_map<std::string, SharedItem<Storage>>;
+
+    DataStoreUpdater(std::shared_ptr<IDataSourceUpdateSink> sink,
+                     std::shared_ptr<IDataStore> store);
+
+    std::unique_ptr<IConnection> OnFlagChange(ChangeHandler handler) override;
+
+    void Init(launchdarkly::data_model::SDKDataSet data_set) override;
+    void Upsert(std::string const& key, FlagDescriptor flag) override;
+    void Upsert(std::string const& key, SegmentDescriptor segment) override;
+    ~DataStoreUpdater() override = default;
+
+    DataStoreUpdater(DataStoreUpdater const& item) = delete;
+    DataStoreUpdater(DataStoreUpdater&& item) = delete;
+    DataStoreUpdater& operator=(DataStoreUpdater const&) = delete;
+    DataStoreUpdater& operator=(DataStoreUpdater&&) = delete;
+
+   private:
+    bool HasListeners() const;
+
+    template <typename FlagOrSegment>
+    void UpsertCommon(
+        DataKind kind,
+        std::string key,
+        SharedItem<FlagOrSegment> existing,
+        launchdarkly::data_model::ItemDescriptor<FlagOrSegment> updated) {
+        if (existing && (updated.version <= existing->version)) {
+            // Out of order update, ignore it.
+            return;
+        }
+
+        dependency_tracker_.UpdateDependencies(key, updated);
+
+        if (HasListeners()) {
+            auto updated_deps = DependencySet();
+            dependency_tracker_.CalculateChanges(kind, key, updated_deps);
+            NotifyChanges(updated_deps);
+        }
+
+        sink_->Upsert(key, updated);
+    }
+
+    template <typename FlagOrSegment>
```

Comment:
Still FlagOrSegment, because DataKind is a type, which is also used by this same method. Which is inconvenient.